sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffleAsyncCoalesceIterator.scala

```diff
@@ -0,0 +1,128 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids
+
+import java.util.concurrent.{Callable, Future}
+
+import ai.rapids.cudf.{NvtxColor, NvtxRange}
+import com.nvidia.spark.rapids.Arm.{closeOnExcept, withResource}
+import com.nvidia.spark.rapids.io.async.{ThrottlingExecutor, TrafficController}
+
+import org.apache.spark.TaskContext
+import org.apache.spark.sql.rapids.execution.TrampolineUtil
+import org.apache.spark.sql.types.DataType
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+
+/**
+ * Similar as GpuShuffleCoalesceIterator, but pulling in host batches asynchronously, to
+ * overlap the host batch reading and the downstream GPU operations.
+ */
+class GpuShuffleAsyncCoalesceIterator(iter: Iterator[CoalescedHostResult],
+    dataTypes: Array[DataType],
+    targetBatchSize: Long,
+    outputBatchesMetric: GpuMetric = NoopMetric,
+    outputRowsMetric: GpuMetric = NoopMetric,
+    asyncReadTimeMetric: GpuMetric = NoopMetric,
+    opTimeMetric: GpuMetric = NoopMetric,
+    readThrottlingMetric: GpuMetric = NoopMetric,
+) extends Iterator[ColumnarBatch] {
+
+  val executor =
+    new ThrottlingExecutor(
+      TrampolineUtil.newDaemonCachedThreadPool(
+        "async shuffle read thread for " + Thread.currentThread().getName, 1, 1),
+      TrafficController.getReadInstance,
+      stat => {
+        readThrottlingMetric.add(stat.accumulatedThrottleTimeNs)
+      }
+    )
+
+  // don't try to call TaskContext.get().taskAttemptId() in the backend thread
+  private val taskAttemptID = Option(TaskContext.get()).
+    map(_.taskAttemptId().toString).getOrElse("unknown")
+
+  private lazy val readCallable = new Callable[CoalescedHostResult]() {
+    // The actual async read, including the host batches read and concatenation in
+    // "HostCoalesceIteratorBase.next()".
+    override def call(): CoalescedHostResult = {
+      val nvRangeName = s"Task ${taskAttemptID} - Async Read Batch (Backend)"
+      withResource(new NvtxRange(nvRangeName, NvtxColor.BLUE)) { _ =>
+        iter.next()
+      }
+    }
+  }
+
+  private var readFutureOpt: Option[Future[CoalescedHostResult]] = None
+
+  override def hasNext(): Boolean = GpuMetric.ns(asyncReadTimeMetric, opTimeMetric) {
+    readFutureOpt.isDefined || {
+      // No async read is running when it comes here, so no need synchronization
+      // when accessing the input iterator. "iter.hasNext" should be lightweight
+      // enough, since it just read in a header which is very small.
```

Comment:
I think we should say that this iterator is not thread safe then in the class header. I'd assume that's the case, but I think it would be nice to comment since we are moving to more and more async. You cannot call `hasNext` from a different thread than next because you don't protect  `readFutureOpt` being set. This future can be set in `next` and it is submitted to the background thread and then the option is set, so we could be in a case where there _is_ async work, since you submitted it, but `readFutureOpt=None` (because we haven't set it yet, and we are not protecting that option).